### PR TITLE
Fix #1096 - "Out of gas issue" from brownie, in test_ve_ocean.py

### LINKED ISSUE #1096: "Out of gas issue" from brownie, in test_ve_ocean.py
NOTE: the test that's failing is skipped, as of commit https://github.com/oceanprotocol/ocean.py/commit/e42efcdb703540a4b209b8cf5706f8bd32b9b8b1. We will un-skip the test once we get it passing (and passing with clean changes)

----

### Overview of issue

To reproduce...

First, open `ocean_lib/models/vetest/test_ve_ocean.py` and comment out the `pytest.mark.skip()` line right before `test_ve_ocean1()`.

Then, have barge running.

Then, in a console, run:
```console
pytest ./ocean_lib/models/vetest/test_ve_ocean.py
```

The failure occurs when [`test_ve_ocean.py`](https://github.com/oceanprotocol/ocean.py/blob/main/ocean_lib/models/vetest/test_ve_ocean.py) makes a smart contract call to [`veOCEAN.vy::create_lock()`](https://github.com/oceanprotocol/contracts/blob/840c835df393917e99caa5dc292c8a47b25704f4/contracts/ve/veOCEAN.vy#L412):
```python
veOCEAN.create_lock(TA, t2, {"from": alice_wallet})
```

And the failure is:
```text
brownie.exceptions.VirtualMachineError: out of gas

# more detailed version
self = VirtualMachineError(ValueError("Gas estimation failed: 'VM Exception while processing transaction: revert Can only loc... in the future'. This transaction will likely revert. If you wish to broadcast, you must set the gas limit manually."))
```

### Implementation of createLock():

From [veOCEAN.vy::createLock()](https://github.com/oceanprotocol/contracts/blob/840c835df393917e99caa5dc292c8a47b25704f4/contracts/ve/veOCEAN.vy#L412).

```python
@external
@nonreentrant('lock')
def create_lock(_value: uint256, _unlock_time: uint256):
    """
    @notice Deposit `_value` tokens for `msg.sender` and lock until `_unlock_time`
    @param _value Amount to deposit
    @param _unlock_time Epoch time when tokens unlock, rounded down to whole weeks
    """
    self.assert_not_contract(msg.sender)
    unlock_time: uint256 = (_unlock_time / WEEK) * WEEK  # Locktime is rounded down to weeks
    _locked: LockedBalance = self.locked[msg.sender]

    assert _value > 0  # dev: need non-zero value
    assert _locked.amount == 0, "Withdraw old tokens first"
    assert unlock_time > block.timestamp, "Can only lock until time in the future"
    assert unlock_time <= block.timestamp + MAXTIME, "Voting lock can be 4 years max"

    self._deposit_for(msg.sender, _value, unlock_time, _locked, CREATE_LOCK_TYPE)
```

### Full traceback

```text

config = {'ADDRESS_FILE': '~/.ocean/ocean-contracts/artifacts/address.json', 'DOWNLOADS_PATH': 'consume-downloads', 'METADATA_CACHE_URI': 'http://172.15.0.5:5000', 'NETWORK_NAME': 'development', ...}
factory_deployer_wallet = <LocalAccount '0xe2DD09d719Da89e5a3D0F2549c7E24566e947260'>, ocean_token = <ocean_lib.models.datatoken.Datatoken object at 0x7f94893137c0>
veOCEAN = <ocean_lib.models.ve_ocean.VeOcean object at 0x7f9488b11960>

    @pytest.mark.unit
    #@pytest.mark.skip(reason="Don't skip, once fixed #1096")
    def test_ve_ocean1(config, factory_deployer_wallet, ocean_token, veOCEAN):
        # inspiration from df-py/util/test/veOcean/test_lock.py
        assert veOCEAN.symbol() == "veOCEAN"
    
        OCEAN = ocean_token
    
        alice_wallet = accounts.add()  # new account avoids "withdraw old tokens first"
        factory_deployer_wallet.transfer(alice_wallet, "1 ether")
    
        TA = to_wei(0.0001)
        OCEAN.mint(alice_wallet.address, TA, {"from": factory_deployer_wallet})
    
        veOCEAN.checkpoint({"from": factory_deployer_wallet})
        OCEAN.approve(veOCEAN.address, TA, {"from": alice_wallet})
    
        t0 = chain.time()
        t1 = t0 // WEEK * WEEK + WEEK  # this is a Thursday, because Jan 1 1970 was
        t2 = t1 + WEEK
        chain.sleep(t1 - t0)
    
        assert OCEAN.balanceOf(alice_wallet.address) != 0
    
>       veOCEAN.create_lock(TA, t2, {"from": alice_wallet})

ocean_lib/models/vetest/test_ve_ocean.py:43: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib/python3.10/site-packages/brownie/network/contract.py:1864: in __call__
    return self.transact(*args)
venv/lib/python3.10/site-packages/brownie/network/contract.py:1737: in transact
    return tx["from"].transfer(
venv/lib/python3.10/site-packages/brownie/network/account.py:682: in transfer
    receipt._raise_if_reverted(exc)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Transaction '0x5f3e8151bbb65d7bf8140597c2571bb3b1ad8adefdb1fd5feb0b4b7fae01c380'>
exc = VirtualMachineError(ValueError({'message': 'VM Exception while processing transaction: out of gas', 'code': -32000, 'd... runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)', 'name': 'o'}}))

    def _raise_if_reverted(self, exc: Any) -> None:
        if self.status or CONFIG.mode == "console":
            return
        if not web3.supports_traces:
            # if traces are not available, do not attempt to determine the revert reason
            raise exc or ValueError("Execution reverted")
    
        if self._dev_revert_msg is None:
            # no revert message and unable to check dev string - have to get trace
            self._expand_trace()
        if self.contract_address:
            source = ""
        elif CONFIG.argv["revert"]:
            source = self._traceback_string()
        else:
            source = self._error_string(1)
            contract = state._find_contract(self.receiver)
            if contract:
                marker = "//" if contract._build["language"] == "Solidity" else "#"
                line = self._traceback_string().split("\n")[-1]
                if marker + " dev: " in line:
                    self._dev_revert_msg = line[line.index(marker) + len(marker) : -5].strip()
    
>       raise exc._with_attr(
            source=source, revert_msg=self._revert_msg, dev_revert_msg=self._dev_revert_msg
        )
E       brownie.exceptions.VirtualMachineError: out of gas

venv/lib/python3.10/site-packages/brownie/network/transaction.py:446: VirtualMachineError
------------------------------------------------------------------------------------------ Captured stdout setup ------------------------------------------------------------------------------------------
Transaction sent: 0x059179f0b3684be000f41ecb798bcb7b2286b4019fb18e2011d14643b5dbd1f0
  Gas price: 1e-09 gwei   Gas limit: 74437   Nonce: 392
  ERC20Template.mint confirmed   Block: 1139   Gas used: 67670 (90.91%)

Transaction sent: 0x037d9b245a3af936f50a5f9313e354e37e30a309720b84a1e10de04d16a73d20
  Gas price: 1e-09 gwei   Gas limit: 57923   Nonce: 393
  ERC20Template.mint confirmed   Block: 1140   Gas used: 52658 (90.91%)

Transaction sent: 0x81cfb5f60fd9043dab3740034812f5dcefe9284e9f26cdf2eb2eda5fc661ab8d
  Gas price: 1e-09 gwei   Gas limit: 57923   Nonce: 394
  ERC20Template.mint confirmed   Block: 1141   Gas used: 52658 (90.91%)

------------------------------------------------------------------------------------------ Captured stdout call -------------------------------------------------------------------------------------------
mnemonic: 'friend decrease rubber morning foster carpet sheriff nuclear assault lake order hat'
Transaction sent: 0xcfb467b0179a538d1e760803d659dbbb279b5b8b6405c2a373bfe3498bc6a480
  Gas price: 1e-09 gwei   Gas limit: 21000   Nonce: 395
  Transaction confirmed   Block: 1142   Gas used: 21000 (100.00%)

Transaction sent: 0xd324e2cab6b170c169b9782746eb22b3fd25e0764020c382e387d252304dd743
  Gas price: 1e-09 gwei   Gas limit: 57884   Nonce: 396
  ERC20Template.mint confirmed   Block: 1143   Gas used: 52622 (90.91%)

Transaction sent: 0x848d1c84618c0ed7bb29dbcecbcf6e25612267972e151405d437381ba2ae89e8
  Gas price: 1e-09 gwei   Gas limit: 94426   Nonce: 397
  veOCEAN.checkpoint confirmed   Block: 1144   Gas used: 85842 (90.91%)

Transaction sent: 0xb5e40639e817fbfb7cb06342e7c5a820f5ed95a3dafaa93ebac88990c4976a86
  Gas price: 1e-09 gwei   Gas limit: 48583   Nonce: 0
  ERC20Template.approve confirmed   Block: 1145   Gas used: 44167 (90.91%)

Transaction sent: 0x5f3e8151bbb65d7bf8140597c2571bb3b1ad8adefdb1fd5feb0b4b7fae01c380
  Gas price: 1e-09 gwei   Gas limit: 410637   Nonce: 1
  veOCEAN.create_lock confirmed (out of gas)   Block: 1146   Gas used: 410637 (100.00%)

========================================================================================= short test summary info =========================================================================================
FAILED ocean_lib/models/vetest/test_ve_ocean.py::test_ve_ocean1 - brownie.exceptions.VirtualMachineError: out of gas
=========================================================================================== 1 failed in 30.83s ============================================================================================

```

### CI failure

Example similar CI failure here: (it's called test1 there).
https://github.com/oceanprotocol/ocean.py/actions/runs/3533298422/jobs/5928743050#step:7:137

It occurred for ve contracts when creating the lock.

### PR DESCRIPTION
**Pull request by mariacarmina**
Fixes #1096  .

Changes proposed in this PR:

- added fix for `out of gas` issue.
- skip remote test if account has no funds.

### CODE REVIEW

File: ocean_lib/models/vetest/test_ve_ocean.py [hunk 1]
```diff
@@ -19,6 +20,9 @@
 @pytest.mark.unit
 def test1(config, factory_deployer_wallet, ocean_token, veOCEAN):
     # inspiration from df-py/util/test/veOcean/test_lock.py
+    brownie.network.gas_limit(chain.block_gas_limit)
```
Comment (trentmc) [inline ocean_lib/models/vetest/test_ve_ocean.py:23]: Using `gas_limit()` is only relevant for pre-1559. [ref - brownie docs on gas](https://eth-brownie.readthedocs.io/en/stable/core-gas.html).

And, ganache has 1559. [ref](https://twitter.com/trufflesuite/status/1440397961150697480).

Therefore you should be able to delete this line.

If it fails unless the line's there, we need to update our mental model I guess?

File: ocean_lib/models/vetest/test_ve_ocean.py [hunk 2]
```diff
@@ -19,6 +20,9 @@
...
+    priority_fee(chain.priority_fee)
```
Comment (trentmc) [inline ocean_lib/models/vetest/test_ve_ocean.py:24]: There are two references to `chain` but I don't see a line like `from brownie.network import chain`. Is it set in a higher-up conftest? Or simply an omission here?

**Review by trentmc — CHANGES_REQUESTED** [review]

File: ocean_lib/models/vetest/test_ve_ocean.py (hunk 1 above)
Comment (mariacarmina) [reply ocean_lib/models/vetest/test_ve_ocean.py:23]: I will delete the gas limit then, thanks.

File: ocean_lib/models/vetest/test_ve_ocean.py (hunk 2 above)
Comment (trentmc) [reply ocean_lib/models/vetest/test_ve_ocean.py:23]: ah yes, I see it now. thanks

**Review by trentmc — APPROVED** [review]

**Comment by trentmc** [comment]
@mariacarmina I think the issue is that - while in theory ganache is supposed to implement 1559, _in practice_ it doesn't! And that's what's causing this new bug.

To test this out: 
- remove the 1559 `priority_fee()` approach
- and instead have a `gas_limit()` approach such as:
```python
from brownie.network import priority_fee
priority_fee("auto") # will dynamically determine the priority fee using web3.eth.max_priority_fee
```

**Comment by mariacarmina** [comment]
> @mariacarmina I think the issue is that - while in theory ganache is supposed to implement 1559, _in practice_ it doesn't! And that's what's causing this new bug.
> 
> To test this out:
> 
>     * remove the 1559 `priority_fee()` approach
> 
>     * and instead have a pre-1559 gas_limit approach. Pick a fast/aggressive one. Details in https://eth-brownie.readthedocs.io/en/stable/core-gas.html
Nice catch! Will implement this, thanks!

**Comment by mariacarmina** [comment]
@trentmc I found [here](https://ethereum.stackexchange.com/questions/94480/out-of-gas-for-a-transaction-on-ganache) a potential answer and I will give it a try. This example is for truffle but I will try to hardcode the gas price for ganache.

**Comment by trentmc** [comment]
Fingers crossed!

One Q that's been nagging at me: why are we seeing issues at all on this? It's ganache, normally we never see issues. What's special about this particular test that causes the problems? cc @mariacarmina @calina-c

**Comment by mariacarmina** [comment]
I have added right now the gas limit to every contract call. I have reproduced the issue locally and that was the fix. If I wanted to set the gas price to `20 gwei` which is the standard value for ganache according to the [official docs ](https://trufflesuite.com/docs/truffle/reference/configuration/) of truffle suite, I got a different error which is the following:
```python
ValueError: sender doesn't have enough funds to send tx. The upfront cost is: 200000000000000000000 and the sender's account only has: 1000000000000000000
```

**Comment by mariacarmina** [comment]
> Fingers crossed!
> 
> One Q that's been nagging at me: why are we seeing issues at all on this? It's ganache, normally we never see issues. What's special about this particular test that causes the problems? cc @mariacarmina @calina-c

The gas limit was not recognized probably, maybe is a good practice to provide it explicitly for further use.

**Comment by trentmc** [comment]
> One Q that's been nagging at me: why are we seeing issues at all on this? It's ganache, normally we never see issues. What's special about this particular test that causes the problems? cc @mariacarmina @calina-c

We need to get to the bottom of this, before we starting interjecting with specialized argument values in each smart contract call.

test_ve_ocean.py isn't trying to do anything special. So what's going on?

**Comment by mariacarmina** [comment]
> > One Q that's been nagging at me: why are we seeing issues at all on this? It's ganache, normally we never see issues. What's special about this particular test that causes the problems? cc @mariacarmina @calina-c
> 
> We need to get to the bottom of this, before we starting interjecting with specialized argument values in each smart contract call.
> 
> test_ve_ocean.py isn't trying to do anything special. So what's going on?

Ok, I will search for better & reliable arguments for this issue.

File: tests/readmes/test_readmes.py [hunk 3]
```diff
@@ -97,6 +98,14 @@ def test_script_execution(script, monkeypatch):
...
 
 
+@pytest.mark.skipif(
+    VirtualMachineError(
+        ValueError(
+            {"code": -32000, "message": "insufficient funds for gas * price + value"}
+        )
+    ),
+    reason="requires more funds for remote accounts",
+)
```
Comment (trentmc) [inline tests/readmes/test_readmes.py:108]: Please undo all the changes in this file, ie to "skip if out of funds".

Why: it's more involved than you think; it's got its own issue that I was working on; and I just merged the fix into main.
- The issue: #1109 
- The PR: https://github.com/oceanprotocol/ocean.py/pull/1111
- The merge commit: https://github.com/oceanprotocol/ocean.py/commit/6985a76a6bf194d4fcf540844b075eec0010e660

**Comment by trentmc** [comment]
Hi @mariacarmina I spent some time at this today, sporadically during my Sunday, via a new PR: https://github.com/oceanprotocol/ocean.py/pull/1115

Two changes:
- I fixed the "out of gas" issues using the tactic that you'd use in this PR:  "gas_limit": chain.
- I also ran into an EVM error about it referring to txs in the future. The fix: make test use block.timestamp, not chain.time, to be in line with smart contract.

In the new PR, it passes locally and in the CI. Therefore I was able to close this issue. And we can close this PR too. Thanks for your work on it!!

**Pull request closed without merging**